sympy/polys/matrices/lll.py

```diff
@@ -0,0 +1,85 @@
+from math import floor as mfloor
+from typing import Tuple
+
+from sympy.polys.domains import ZZ, QQ
+from sympy.polys.matrices.exceptions import DMRankError, DMShapeError, DMValueError, DMDomainError
+
+logic_error_template = '''FAILURE IN ALGORITHM LOGIC!!!!
+Please file a bug report at https://github.com/sympy/sympy and include this input:
+{}'''
+linear_dependent_error = "input matrix contains linearly dependent rows"
+
+
+def ddm_lll(x, delta=QQ(3, 4)):
+    if QQ(1, 4) >= delta or delta > QQ(1, 1):
+        DMValueError("delta must lie in range (0.25, 1)")
+    if x.shape[0] > x.shape[1]:
+        DMShapeError("input matrix must have shape (m, n) with m <= n")
+    if x.domain != ZZ:
+        DMDomainError("input matrix domain must be ZZ")
+    m = x.shape[0]
+    n = x.shape[1]
+    k = 1
+    y = x.copy()
+    y_star = x.zeros((m, n), QQ)
+    mu = x.zeros((m, m), QQ)
+    g_star = [QQ(0, 1) for _ in range(m)]
+    half = QQ(1, 2)
+
+    def closest_integer(x):
+        return ZZ(mfloor(x + half))
+
+    def lovasz_condition(k: int) -> bool:
+        return g_star[k] >= ((delta - pow(mu[k][k - 1], 2)) * g_star[k - 1])
+
+    def mu_small(k: int, j: int) -> bool:
+        return abs(mu[k][j]) <= half
+
+    def dot_rows(x, y, rows: Tuple[int, int]):
+        return sum([x[rows[0]][z] * y[rows[1]][z] for z in range(x.shape[1])])
+
+    def reduce_row(mu, y, rows: Tuple[int, int]):
+        r = closest_integer(mu[rows[0]][rows[1]])
+        y[rows[0]] = [y[rows[0]][z] - r * y[rows[1]][z] for z in range(n)]
+        for j in range(rows[1]):
+            mu[rows[0]][j] -= r * mu[rows[1]][j]
+        mu[rows[0]][rows[1]] -= r
+
+    for i in range(m):
+        y_star[i] = [QQ.convert_from(z, ZZ) for z in y[i]]
+        for j in range(i):
+            try:
+                mu[i][j] = dot_rows(y, y_star, (i, j)) / g_star[j]
+            except ZeroDivisionError:
+                raise DMRankError(linear_dependent_error)
+            y_star[i] = [y_star[i][z] - mu[i][j] * y_star[j][z] for z in range(n)]
+        g_star[i] = dot_rows(y_star, y_star, (i, i))
+    while k < m:
+        if not mu_small(k, k - 1):
+            reduce_row(mu, y, (k, k - 1))
+        if lovasz_condition(k):
+            for l in range(k - 2, -1, -1):
+                if not mu_small(k, l):
+                    reduce_row(mu, y, (k, l))
+            k += 1
+        else:
+            nu = mu[k][k - 1]
+            alpha = g_star[k] + pow(nu, 2) * g_star[k - 1]
+            try:
+                beta = g_star[k - 1] / alpha
+            except ZeroDivisionError:
+                raise DMRankError(linear_dependent_error)
+            mu[k][k - 1] = nu * beta
+            g_star[k] = g_star[k] * beta
+            g_star[k - 1] = alpha
+            y[k], y[k - 1] = y[k - 1], y[k]
+            for j in range(k - 1):
+                mu[k][j], mu[k - 1][j] = mu[k - 1][j], mu[k][j]
```

Comment:
If  `mu` is a list of lists then is this just swapping rows?

If so then rather than a loop it's just:
```
mu[k], mu[k - 1] = mu[k - 1], mu[k]
```
That just swaps the list rather than swapping the items in the lists individually.

Reply:
In this case it isn't the full row we are swapping. The j only indexes up to k - 2, so it is a portion of a rows we swap.

Reply:
I guess this is okay but it's more efficient to use slicing in that case:
```
mu[k][:k - 1], mu[k - 1][:k - 1] = mu[k - 1][:k - 1], mu[k][:k - 1]
```
Basically in Python explicit loops are usually slow compared to using builtin methods when they are available.